File: algorithms/sorting/counting-sort/python/counting_sort.py
def counting_sort(arr: list[int]) -> list[int]:
    """
    Counting Sort implementation.
    Efficient for sorting integers with a known small range.
    """
    if not arr:
        return []

    min_val = min(arr)
    max_val = max(arr)
    range_val = max_val - min_val + 1

    count = [0] * range_val
    output = [0] * len(arr)

    # Store count of each character
    for i in range(len(arr)):
        count[arr[i] - min_val] += 1

    # Change count[i] so that count[i] now contains actual
    # position of this character in output array
    for i in range(1, len(count)):
        count[i] += count[i - 1]

    # Build the output character array
    for i in range(len(arr) - 1, -1, -1):
        output[count[arr[i] - min_val] - 1] = arr[i]
        count[arr[i] - min_val] -= 1

    return output

File: algorithms/sorting/counting-sort/kotlin/CountingSort.kt
package algorithms.sorting.counting

/**
 * Counting Sort implementation.
 * Efficient for sorting integers with a known small range.
 */
object CountingSort {
    fun sort(arr: IntArray): IntArray {
        if (arr.isEmpty()) {
            return IntArray(0)
        }

        var min = arr[0]
        var max = arr[0]

        for (i in 1 until arr.size) {
            if (arr[i] < min) min = arr[i]
            if (arr[i] > max) max = arr[i]
        }

        val range = max - min + 1
        val count = IntArray(range)
        val output = IntArray(arr.size)

        for (x in arr) {
            count[x - min]++
        }

        for (i in 1 until count.size) {
            count[i] += count[i - 1]
        }

        for (i in arr.size - 1 downTo 0) {
            output[count[arr[i] - min] - 1] = arr[i]
            count[arr[i] - min]--
        }

        return output
    }
}
